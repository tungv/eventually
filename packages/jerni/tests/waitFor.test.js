const JerniPersistenceTimeout = require("../lib/JerniPersistenceTimeout");
const createJourney = require("../lib/createJourney2");
const makeServer = require("./makeServer");
const makeTestStore = require("./makeTestStore");
const makeTestLogger = require("./makeTestLogger");

describe("#waitFor", () => {
  it("should throw JerniPersistenceTimeout error after a given amount of time to wait", async () => {
    jest.setTimeout(1000);
    const { server } = await makeServer({
      port: 19060,
      ns: "test_wait_for",
    });

    try {
      const store = makeTestStore(event => event.id);
      const db = await store.getDriver();

      const journey = createJourney({
        writeTo: "http://localhost:19060",
        stores: [store],
      });

      const event = await journey.commit({ type: "test" });

      await expect(journey.waitFor(event, 100)).rejects.toBeInstanceOf(
        JerniPersistenceTimeout,
      );
      expect(db).toHaveLength(0);
    } finally {
      server.close();
    }
  });

  it("should wait until the event is fully committed to stores", async () => {
    jest.setTimeout(1000);
    const [logger1, appLogs] = makeTestLogger();
    const [logger2, jobLogs] = makeTestLogger();
    const { server } = await makeServer({
      port: 19060,
      ns: "test_wait_for",
    });

    try {
      const store = makeTestStore(event => event.id);
      const journey = createJourney({
        writeTo: "http://localhost:19060",
        stores: [store],
        logger: logger1,
      });

      const subInstance = createJourney({
        writeTo: "http://localhost:19060",
        stores: [store],
        logger: logger2,
      });
      const db = await store.getDriver();

      (async function() {
        // eslint-disable-next-line no-unused-vars
        for await (const item of subInstance.begin({ pulseTime: 1 })) {
        }
      })();

      await sleep(100);
      const event = await journey.commit({ type: "test" });
      await journey.waitFor(event, 500);
      expect(db).toHaveLength(1);
    } finally {
      server.destroy();
      expect(appLogs.join("\n")).toMatchSnapshot("application logs");
      expect(jobLogs.join("\n")).toMatchSnapshot("jerni cli logs");
    }
  });
});

const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));
