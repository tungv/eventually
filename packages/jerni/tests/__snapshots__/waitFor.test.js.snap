// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#waitFor should wait until the event is fully committed to stores: application logs 1`] = `""`;

exports[`#waitFor should wait until the event is fully committed to stores: jerni cli logs 1`] = `
"[DBG] journey.begin({{
  pulseCount: 200,
  pulseTime: 1,
  serverUrl: undefined,
  cleanStart: false
}})
[DBG] sending http request to: http://localhost:19060/subscribe
[DBG] headers { 'last-event-id': '0', includes: '' }
[DBG] socket opened!
[INF] start receiving data
[DBG] handling event #1
[DBG] done"
`;
